### ISSUE TITLE
Feature: custom rpc method `eth_getBlocksFromArray` for Anvil

### ISSUE THREAD

**Issue by sambacha** [foundry-rs/foundry#2405]
### Component

Anvil

### `eth_getBlocksFromArray` 

Get blocks with logs in a batch RPC request with optional consistency.
uses `eth_getBlockByNumber` and `eth_getLogs`

<table>
<tr>
	<td> param {Number[]}  
	<td>  `blockNumbers` Array of block numbers to index.
<tr>
	<td> {Promise<Object[]>}   
	<td>  Blocks with logs.
</table>

### Additional context

part of https://github.com/foundry-rs/foundry/issues/2396

**Comment by mattsse** [comment]
so basically `eth_getBlockByNumber` but for a range?

**Comment by sambacha** [comment]
> so basically `eth_getBlockByNumber` but for a range?

yes

**Comment by zemse** [comment]
Hey @sambacha can you pls clarify the feature? Is the result simply `Vec<Block<Transaction>>` (result of `eth_getBlockByNumber(_, true)`) or how should it include logs?

**Comment by sambacha** [comment]
> Hey @sambacha can you pls clarify the feature? Is the result simply `Vec<Block<Transaction>>` (result of `eth_getBlockByNumber(_, true)`) or how should it include logs?

```typescript
Eth.prototype.getBlocksFromArray = async function getBlocksFromArray(blockNumbers) {
    let batch = [];

    for (let bn of blockNumbers) {
        batch.push(this.client.request('eth_getBlockByNumber', ['0x' + bn.toString(16), true], undefined, false));
        batch.push(this.client.request('eth_getLogs', [formatLogOptions(bn, bn)], undefined, false));
    }

    let blocks = [];

    const responses = await this.request(batch);

    for (let i = 0; i < responses.length - 1; i += 2) {
        const block = responses[i].result;

        blocks.push(Object.assign(block, {
            number:    parseInt16(block.number),
            timestamp: parseInt16(block.timestamp),
            logs:      responses[i + 1].result
        }));
    }

    return blocks;
};
```

**Comment by onbjerg** [comment]
I asked in the other issue related to a new RPC method as well, but what is the primary use case of this? No tooling supports it, no other node has it (I think?) and we don't have a use case for it in Foundry

**Comment by tynes** [comment]
Can't you use a rpc batch provider for this functionality?  Adding new RPC methods that are not part of the official Ethereum RPC spec creates a maintenance burden

**Comment by sambacha** [comment]
> I asked in the other issue related to a new RPC method as well, but what is the primary use case of this? No tooling supports it, no other node has it (I think?) and we don't have a use case for it in Foundry

eth_getBatchBlockDataByNumbers new rpc function that takes one param (array of block numbers) and returns an object with both the minimal block data and logs, ({<blockNo>: {number: <blockNo>, hash: <blockHash>, parentHash: <blockParentHash>, difficulty: <blockDifficulty>, logs: <CompactLog[]>})

**Comment by sambacha** [comment]
> Can't you use a rpc batch provider for this functionality? Adding new RPC methods that are not part of the official Ethereum RPC spec creates a maintenance burden

ofc you could. However the RPC method works behind the scenes is up to the implementation. 


There are tons of non-standard RPC methods that are not a part of the official RPC spec. A plugin system would solve this, so thx

**Comment by sambacha** [comment]
> Can't you use a rpc batch provider for this functionality? Adding new RPC methods that are not part of the official Ethereum RPC spec creates a maintenance burden

https://github.com/foundry-rs/foundry/issues/2446

**Comment by onbjerg** [comment]
I understand what the request does, I just don't really see why the additional maintenance burden (which is not just code, it's support and docs too) is "worth it" for something that can be solved by existing calls (e.g. via batching). All of the custom RPC methods I've seen in other nodes have pretty much always added functionality that is not solveable with existing methods (e.g. tracing)

**Comment by onbjerg** [comment]
Closed as per comments in #2425
